### ISSUE TITLE
Assert custom function taking a Plug.Conn

### ISSUE THREAD

**Issue by waiting-for-dev** [boydm/phoenix_integration#10]
Hi!

I find this library very elegant and handy to make integration test readable without compromising simplicity and speed. Congratulations!

Today I came across a not covered scenario that maybe could be worthwhile to add to it. The idea is to allow asserting custom functions that take `Plug.Conn` as first parameter.

I describe my scenario: I'm testing that a user signs in. With `assert_response` there is no problem to assert that the name of the user is printed on screen, but I would like to really be sure that the user has been added to the session (or refute it, in an invalid attempt). As I'm using [Guardian](https://github.com/ueberauth/guardian) for that, now I have something like the following:

```elixir
#...
  defp assert_user_in_session(conn) do
    assert Guardian.Plug.current_resource(conn)
    conn
  end

  defp refute_user_in_session(conn) do
    refute Guardian.Plug.current_resource(conn)
    conn
  end

  describe "when credentials are valid" do
    test "signs the user in", %{conn: conn} do
      get(conn, session_path(conn, :new))
      # sign_in is just a helper function that uses follow_form
      |> sign_in("some@email.com", "secure_password")
      |> assert_user_in_session
      |> assert_response(
        status: 200,
        html: "some@email.com"
      )
    end
  end
#...
```

Adding the functionality to this library, we could write something like:

```elixir
      get(conn, session_path(conn, :new))
      |> sign_in("some@email.com", "secure_password")
      |> assert_response(
        status: 200,
        html: "some@email.com",
        # Ideally, accepting several `function` keys
        function: fn(conn) -> Guardian.Plug.current_resource(conn) end
      )
```

or

```elixir
      get(conn, session_path(conn, :new))
      |> sign_in("some@email.com", "secure_password")
      |> assert_response(
        status: 200,
        html: "some@email.com",
      )
      |> assert_function(
        fn(conn) -> Guardian.Plug.current_resource(conn) end
      )
```

What do you think? Maybe another noun other than `function` could be used for the key/function name.

**Comment by boydm** [comment]
This is a good idea.

Let me think about this for a bit to form an opinion on the right way to model it. Both the function: parameter and the assert_function have merit and I've already gone back and forth on which I like better.

**Comment by waiting-for-dev** [comment]
Thanks for your feedback @boydm .

Sure, there is no rush for this. Let me know if you need something, please.

**Comment by boydm** [comment]
Just checking here...
Are you suggesting that the assertion is that the function returns something truthy?
Or are you planning to write more assertions inside the callback?

**Comment by boydm** [comment]
I went with adding a :value type to assert_response. Try it out and let me know how it works for you before I publish it up to hex.

You can pull this version by updating your mix file deps like this.
{:phoenix_integration, git: "https://github.com/boydm/phoenix_integration.git"}

after I publish it to hex, you should revert to the normal way of getting deps, but with the new version number...

**Comment by waiting-for-dev** [comment]
Hey @boydm . I have already tested it and it looks great :+1: I also think that it is better having it integrated to `assert_response` and `refute_response` as a new key.

Thanks for your work on this!

**Comment by boydm** [comment]
It is published and pushed to Hex as 0.3.0
Thank you.
